chainerx_cc/chainerx/routines/linalg.cc

```diff
@@ -25,17 +27,33 @@ Array Dot(const Array& a, const Array& b) {
     // TODO(beam2d): dtype conversion
     CheckEqual(a.dtype(), b.dtype());
 
-    // TODO(beam2d): Support it. Need to transpose b so that the inner-product axis is moved to the top.
-    if (b.ndim() > 2) {
-        throw NotImplementedError{"dot does not support rhs operand with ndim > 2"};
-    }
+    Array modified_b;
 
     Shape out_shape{};
-    std::copy(a.shape().begin(), a.shape().end() - 1, std::back_inserter(out_shape));
-    std::copy(b.shape().begin() + 1, b.shape().end(), std::back_inserter(out_shape));
+
+    if (b.ndim() > 2) {
+        std::vector<int> b_transpose_axes{};
+        std::vector<int> axes_index(b.ndim());
+        std::iota(axes_index.begin(), axes_index.end(), 0);
+        std::copy(axes_index.begin(), axes_index.end()-2, std::back_inserter(b_transpose_axes));
+        std::reverse_copy(axes_index.end()-2, axes_index.end(), std::back_inserter(b_transpose_axes));
+
+        Axes axes(b_transpose_axes.begin(), b_transpose_axes.end());
+        modified_b = b.Transpose(axes);
+
+        std::copy(a.shape().begin(), a.shape().end()-1, std::back_inserter(out_shape));
+        std::copy(modified_b.shape().begin(), modified_b.shape().end()-1, std::back_inserter(out_shape));
+
+        modified_b = modified_b.Reshape({-1, modified_b.shape()[modified_b.ndim()-1]});
```

Comment:
You can probably use `modified_b.shape().back()` here https://github.com/chainer/chainer/blob/master/chainerx_cc/chainerx/stack_vector.h#L137.